src/librustc_metadata/rmeta/encoder.rs

```diff
@@ -163,26 +165,29 @@ impl<'a, 'tcx> Encodable<EncodeContext<'a, 'tcx>> for ExpnId {
 
 impl<'a, 'tcx> Encodable<EncodeContext<'a, 'tcx>> for Span {
     fn encode(&self, s: &mut EncodeContext<'a, 'tcx>) -> opaque::EncodeResult {
-        if self.is_dummy() {
-            return TAG_INVALID_SPAN.encode(s);
-        }
-
-        let span = self.data();
+        let mut span = self.data();
 
         // The Span infrastructure should make sure that this invariant holds:
         debug_assert!(span.lo <= span.hi);
 
-        if !s.source_file_cache.0.contains(span.lo) {
+        // Updates `source_file_cache` to point to the `SourceFile` containing
+        // `data.lo`
+        let update_source_file_cache = |s: &mut EncodeContext<'a, 'tcx>, data: &SpanData| {
             let source_map = s.tcx.sess.source_map();
-            let source_file_index = source_map.lookup_source_file_idx(span.lo);
+            let source_file_index = source_map.lookup_source_file_idx(data.lo);
             s.source_file_cache =
                 (source_map.files()[source_file_index].clone(), source_file_index);
+        };
+
+        if !s.source_file_cache.0.contains(span.lo) {
+            update_source_file_cache(s, &span);
         }
 
         if !s.source_file_cache.0.contains(span.hi) {
             // Unfortunately, macro expansion still sometimes generates Spans
             // that malformed in this way.
-            return TAG_INVALID_SPAN.encode(s);
+            span = DUMMY_SP.data();
+            update_source_file_cache(s, &span);
```

Comment:
This case *could* be common.
Not sure, you'll probably have to collect some statistics.